### PR TITLE
[35815] Add visible show page for all users to groups

### PR DESCRIPTION
**Pull request by oliverguenther**
To that end, move the groups routes out of /admin. This also removes the `XML` responses which are not documented nor used anywhere to my knowledge.

https://community.openproject.com/wp/35815

### CODE REVIEW

File: modules/costs/lib/costs/engine.rb [hunk 1]
```diff
@@ -220,11 +220,11 @@ class Engine < ::Rails::Engine
...
       # format and not a string like "10 EUR"
-      schema :overall_costs,
-             type: 'String',
+      schema :sachnummer,
+             type: 'Sachnummer',
              required: false,
-             writable: false,
-             show_if: ->(*) { represented.project && represented.project.costs_enabled? }
+             writable: true,
+             show_if: ->(*) { represented.project && euerPluginIstAktiv }
```
Comment (ulferts) [inline modules/costs/lib/costs/engine.rb]: This has probably creeped in from a a different activity.

File: spec/features/groups/group_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    end
+  end
+
+  context 'as a regular user' do
+    let(:current_user) { FactoryBot.create :user }
+
+    scenario 'I can visit the group page' do
+      visit group_path(group)
+      expect(page).to have_selector('h2', text: "Bob's Team")
+      expect(page).to have_no_selector('.toolbar-item')
```
Comment (ulferts) [inline spec/features/groups/group_show_spec.rb:57]: Ideally, this would also test for the name of a group member not being displayed.

Comment (ulferts) [reply spec/features/groups/group_show_spec.rb:57]: Or, we could display it for all those users having the manage_members permission.

File: config/routes.rb [hunk 3]
```diff
@@ -465,6 +453,18 @@
...
+    member do
+      # this should be put into it's own resource
+      match '/members' => 'groups#add_users', via: :post, as: 'members_of'
+      match '/members/:user_id' => 'groups#remove_user', via: :delete, as: 'member_of'
+      # this should be put into it's own resource
+      match '/memberships/:membership_id' => 'groups#edit_membership', via: :put, as: 'membership_of'
+      match '/memberships/:membership_id' => 'groups#destroy_membership', via: :delete
+      match '/memberships' => 'groups#create_memberships', via: :post, as: 'memberships_of'
+    end
+  end
```
Comment (ulferts) [inline config/routes.rb]: Can we get the show page to be outside of the admin scope but the rest to stay inside it? How much of a hassle would that be?

**Review by ulferts — CHANGES_REQUESTED** [review]
Mostly looking good. Thanks for transferring the legacy specs.

File: modules/costs/lib/costs/engine.rb (hunk 1 above)
Comment (oliverguenther) [reply modules/costs/lib/costs/engine.rb]: Haha nice :laughing: This was an example I made in a screenshare with amg

File: config/routes.rb (hunk 3 above)
Comment (oliverguenther) [reply config/routes.rb]: That's easily doable with two `resources` blocks. Will do. /edit This however results in two different path helpers that have to be used

File: spec/features/groups/group_show_spec.rb (hunk 2 above)
Comment (oliverguenther) [reply spec/features/groups/group_show_spec.rb:57]: I changed the check to match the response for the groups API , which shows members when a user has the `manage_members` or `view_members` permission in a project the group is part of. https://github.com/opf/openproject/pull/8949/files#diff-af3ad490ca40c9ba7d5641ead932a1b6af38ec2b4af21c0f8468b11ed97f0ba5R156-R159

**Comment by oliverguenther** [comment]
I believe I addressed all your remarks

**Review by ulferts — APPROVED** [review]
:+1:

**Comment by maxemann96** [comment]
Thanks for the constant evolution of OpenProject. Unfortunately we used this "hidden xml feature" as a possibility to maintain our OpenProject Groups automatically since the API v3 does only support listing a group with known id. As this is removed are there any plans to extend the Api v3 Groups in the near future?

**Comment by ulferts** [comment]
Hi @maxemann96, what information do you need from the API? Does it need to include the group members or do you simply need the names. 

The API v3 in fact already provides an endpoint for listing all principals (groups and users) which can then be filtered to only return the groups:
```
/api/v3/principals?filters=[{"type":{"operator":"=", "values": ["Group"]}}]
```
Admittedly, as of now the group members are not listed for the group resource and there is no date yet on when that will be implemented.

https://docs.openproject.org/api/endpoints/principals/

**Comment by maxemann96** [comment]
Hi @ulferts 
thanks for your fast answer. Our use case is to have CRUD operations on groups and List/Add/Remove Members from/to group. Currently we are doing this with the xml Feature. We figured out, that we can also list people in a group with
`/api/v3/users?filters=[{"group":{"operator":"=","values":[$GROUP_ID]}}]`

As a result the following operations we are doing currently with the xml api.

- Group Create
- Group Delete
- Add Group Member
- Remove Group Member

This change is a huge regression for us, as we now need to parse the html page, extract the csrf token and the success/error message of the Operation.

**Comment by ulferts** [comment]
Hi @maxemann96,

I have updated our ticket to track that change: https://community.openproject.org/projects/openproject/work_packages/18812 but I cannot give you a date on when that will be implemented. I have added your request as a vote to the feature but compared to other feature requests, this is not on the top of the list.

I'm sorry that we removed the functionality for you to manage groups programatically. The 'old' xml based approach would require extra effort for us to maintain and we will focus on the API v3 instead.

**Comment by maxemann96** [comment]
> Hi @maxemann96, what information do you need from the API? Does it need to include the group members or do you simply need the names.
> 
> The API v3 in fact already provides an endpoint for listing all principals (groups and users) which can then be filtered to only return the groups:
> 
> ```
> /api/v3/principals?filters=[{"type":{"operator":"=", "values": ["Group"]}}]
> ```
> 
> Admittedly, as of now the group members are not listed for the group resource and there is no date yet on when that will be implemented.
> 
> https://docs.openproject.org/api/endpoints/principals/

The principals endpoint lists only groups/users which are assigned to any project. As a result the endpoint cannot be used to get all group names, since a newly created group is not assigned to any project or did I miss something?

**Pull request merged**